src/gfn/samplers.py

```diff
@@ -264,3 +263,324 @@ def sample_trajectories(
+        if back_steps is None:
+            assert (
+                back_ratio is not None
+            ), "Either kwarg `back_steps` or `back_ratio` must be specified"
+            K = torch.ceil(back_ratio * (trajectories.when_is_done - 1)).long()
+        else:
+            K = torch.where(
+                back_steps > trajectories.when_is_done,
+                trajectories.when_is_done,
+                back_steps,
+            )
+
+        backward_trajectories = self.backward_sampler.sample_trajectories(
+            env,
+            states=trajectories.last_states,
+            conditioning=conditioning,
+            save_estimator_outputs=save_estimator_outputs,
+            save_logprobs=save_logprobs,
+            **policy_kwargs,
+        )
+        # Calculate the forward probability if needed (metropolis-hastings).
+        ### COPIED FROM `TrajectoryBasedGFlowNet.get_pfs_and_pbs` ###
+        if use_metropolis_hastings:
+            ### FIXME: I realize that the trajectory needs to be reverted to get the forward probability.
+            ### TODO: Resolve the issue first https://github.com/GFNOrg/torchgfn/issues/109
+            valid_states = backward_trajectories.states[
+                ~backward_trajectories.states.is_sink_state
+            ]
+            valid_actions = backward_trajectories.actions[
+                ~backward_trajectories.actions.is_dummy
+            ]
+
+            if backward_trajectories.conditioning is not None:
+                cond_dim = (-1,) * len(backward_trajectories.conditioning.shape)
+                traj_len = backward_trajectories.states.tensor.shape[0]
+                masked_cond = backward_trajectories.conditioning.unsqueeze(0).expand(
+                    (traj_len,) + cond_dim
+                )[~backward_trajectories.states.is_sink_state]
+
+                # Here, we pass all valid states, i.e., non-sink states.
+                with has_conditioning_exception_handler("pf", self.estimator):
+                    estimator_outputs = self.estimator(valid_states, masked_cond)
+            else:
+                # Here, we pass all valid states, i.e., non-sink states.
+                with no_conditioning_exception_handler("pf", self.estimator):
+                    estimator_outputs = self.estimator(valid_states)
+
+            # Calculates the log PF of the actions sampled off policy.
+            valid_log_pf_actions = self.estimator.to_probability_distribution(
+                valid_states, estimator_outputs
+            ).log_prob(
+                valid_actions.tensor
+            )  # Using the actions sampled off-policy.
+            log_pf_backward_trajectories = torch.full_like(
+                backward_trajectories.actions.tensor[..., 0],
+                fill_value=0.0,
+                dtype=torch.float,
+            )
+            log_pf_backward_trajectories[
+                ~backward_trajectories.actions.is_dummy
+            ] = valid_log_pf_actions
```

Comment:
This part is copied from `TrajectoryBasedGFlowNet.get_pfs_and_pbs`.

I couldn't find any other way to calculate the `log_pfs` or `log_pbs`, but this is too ugly and too much redundancies.

We need more modular functionality to minimize the redundancy.